server/channels/app/web_broadcast_hooks.go

```diff
@@ -0,0 +1,106 @@
+// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
+// See LICENSE.txt for license information.
+
+package app
+
+import (
+	"fmt"
+
+	"github.com/mattermost/mattermost/server/public/model"
+	pUtils "github.com/mattermost/mattermost/server/public/utils"
+	"github.com/mattermost/mattermost/server/v8/channels/app/platform"
+)
+
+const (
+	broadcastAddMentions  = "add_mentions"
+	broadcastAddFollowers = "add_followers"
+)
+
+func (s *Server) makeBroadcastHooks() map[string]platform.BroadcastHook {
+	return map[string]platform.BroadcastHook{
+		broadcastAddMentions:  &addMentionsBroadcastHook{},
+		broadcastAddFollowers: &addFollowersBroadcastHook{},
+	}
+}
+
+type addMentionsBroadcastHook struct{}
+
+func (h *addMentionsBroadcastHook) ShouldProcess(msg *model.WebSocketEvent, webConn *platform.WebConn, args map[string]any) (bool, error) {
+	if msg.EventType() != model.WebsocketEventPosted {
+		return false, nil
+	}
+
+	mentions, ok := args["mentions"].(model.StringArray)
+	if !ok {
+		return false, fmt.Errorf("Invalid mentions value passed to addMentionsBroadcastHook: %v", args["mentions"])
+	}
```

Comment:
This doesn't work cross-cluster because once the message is unmarshalled from a byte slice, it loses any type info. So mention parsing fails in a cluster scenario with:

```
error [2023-10-16 05:50:30.901 Z] runBroadcastHooks: Encountered error running ShouldProcess for broadcast hook caller="platform/web_broadcast_hook.go:37" error="Invalid mentions value passed to addMentionsBroad
castHook: [q1gtsishkfr388sa19o1iitspr]"
```

To fix this, we need to marshal and unmarshal again like:

```diff
diff --git a/server/channels/app/web_broadcast_hooks.go b/server/channels/app/web_broadcast_hooks.go
index 7c6a51c4ab..77810db878 100644
--- a/server/channels/app/web_broadcast_hooks.go
+++ b/server/channels/app/web_broadcast_hooks.go
@@ -4,6 +4,7 @@
 package app
 
 import (
+	"encoding/json"
 	"fmt"
 
 	"github.com/mattermost/mattermost/server/public/model"
@@ -30,24 +31,46 @@ func (h *addMentionsBroadcastHook) ShouldProcess(msg *model.WebSocketEvent, webC
 		return false, nil
 	}
 
-	mentions, ok := args["mentions"].(model.StringArray)
-	if !ok {
-		return false, fmt.Errorf("Invalid mentions value passed to addMentionsBroadcastHook: %v", args["mentions"])
+	if args["mentions"] == nil {
+		return false, nil
+	}
+
+	buf, err := json.Marshal(args["mentions"])
+	if err != nil {
+		return false, err
 	}
 
+	var mentionArray model.StringArray
+	if err = json.Unmarshal(buf, &mentionArray); err != nil {
+		return false, err
+	}
+
+	fmt.Println("PASSED---------", mentionArray)
+
 	// This hook will only modify the event if the current user is mentioned by the post
-	return pUtils.Contains[string](mentions, webConn.UserId), nil
+	return pUtils.Contains[string](mentionArray, webConn.UserId), nil
 }
 
 func (h *addMentionsBroadcastHook) Process(msg *model.WebSocketEvent, webConn *platform.WebConn, args map[string]any) (*model.WebSocketEvent, error) {
-	mentions, ok := args["mentions"].(model.StringArray)
-	if !ok {
-		return msg, fmt.Errorf("Invalid mentions value passed to addMentionsBroadcastHook: %v", args["mentions"])
+	if args["mentions"] == nil {
+		return nil, nil
+	}
+
+	buf, err := json.Marshal(args["mentions"])
+	if err != nil {
+		return nil, err
 	}
 
+	var mentionArray model.StringArray
+	if err = json.Unmarshal(buf, &mentionArray); err != nil {
+		return nil, err
+	}
+
+	fmt.Println("PASSED---------", mentionArray)
 	hasMention := false
-	if len(mentions) > 0 {
-		hasMention = pUtils.Contains(mentions, webConn.UserId)
+
+	if len(mentionArray) > 0 {
+		hasMention = pUtils.Contains(mentionArray, webConn.UserId)
 	}
 
 	if hasMention {

```

Reply:
Oh, that's kind of gross I tested locally, but that must've been before I checked `ok` since it would've just failed silently then.

I wish there was a better way to reinterpret the type of that so hopefully it won't add too much overhead to use JSON to reinterpret the type. I'll make that change, thanks

Reply:
I put this logic in a helper since I ended up repeating myself. It also has the nice benefit of letting us include skip the marshaling/unmarshaling step if the value is already the correct type which is always the case for single server setups.